Cubecraft doesn't work

```
panic: runtime error: slice bounds out of range

goroutine 34 [running]:
github.com/jhead/phantom/internal/proxy.(*ProxyServer).processDataFromServer(0xc000092000, 0xc0000e4000, 0x53ee60, 0xc0000b8030)
        /Users/justin/dev/phantom/internal/proxy/proxy.go:254 +0x48f
created by github.com/jhead/phantom/internal/proxy.getServerConnection
        /Users/justin/dev/phantom/internal/proxy/proxy.go:213 +0x280
```

--- a/go.mod
+++ b/go.mod
@@ -2,4 +2,7 @@ module github.com/jhead/phantom
 
 go 1.12
 
-require github.com/libp2p/go-reuseport v0.0.1
+require (
+	github.com/libp2p/go-reuseport v0.0.1
+	github.com/stretchr/testify v1.3.0
+)

--- a/internal/proto/proto.go
+++ b/internal/proto/proto.go
@@ -93,6 +93,7 @@ func (r UnconnectedReply) Build() bytes.Buffer {
 	return outBuffer
 }
 
+// Reads pong data from the string off the wire into an empty PongData struct
 func readPong(raw string) PongData {
 	pong := PongData{}
 	pongParts := []interface{}{}
@@ -108,6 +109,8 @@ func readPong(raw string) PongData {
 	return pong
 }
 
+// Turns a PongData into a string that complies with the Bedrock protocol,
+// separating the fields with ;
 func writePong(pong PongData) string {
 	var pongDataFields []string
 	pongDataFieldsRaw := util.MapStructToFields(&pong)

--- a/internal/proto/proto.go
+++ b/internal/proto/proto.go
@@ -3,15 +3,35 @@ package proto
 import (
 	"bytes"
 	"encoding/binary"
+	"fmt"
+	"strings"
+
+	"github.com/jhead/phantom/internal/util"
 )
 
 var UnconnectedReplyID byte = 0x1C
 
 type UnconnectedReply struct {
-	PingTime   []byte
-	ID         []byte
-	Magic      []byte
-	ServerName string
+	PingTime []byte
+	ID       []byte
+	Magic    []byte
+	Pong     PongData
+}
+
+type PongData struct {
+	Edition         string
+	MOTD            string
+	ProtocolVersion string
+	Version         string
+	Players         string
+	MaxPlayers      string
+	ServerID        string
+	SubMOTD         string
+	GameType        string
+	NintendoLimited string
+	// Specifically omit these two because they cause issues
+	// Port4           string
+	// Port6           string
 }
 
 func ReadUnconnectedReply(in []byte) (reply *UnconnectedReply, err error) {
@@ -36,19 +56,19 @@ func ReadUnconnectedReply(in []byte) (reply *UnconnectedReply, err error) {
 		return nil, err
 	}
 
-	serverNameLenBytes := make([]byte, 2)
-	if _, err := buf.Read(serverNameLenBytes); err != nil {
+	pongLenBytes := make([]byte, 2)
+	if _, err := buf.Read(pongLenBytes); err != nil {
 		return nil, err
 	}
 
-	serverNameLen := binary.BigEndian.Uint16(serverNameLenBytes)
+	pongLen := binary.BigEndian.Uint16(pongLenBytes)
 
-	serverNameBytes := make([]byte, serverNameLen)
-	if _, err := buf.Read(serverNameBytes); err != nil {
+	pongDataBytes := make([]byte, pongLen)
+	if _, err := buf.Read(pongDataBytes); err != nil {
 		return nil, err
 	}
 
-	reply.ServerName = string(serverNameBytes)
+	reply.Pong = readPong(string(pongDataBytes))
 
 	return
 }
@@ -61,12 +81,39 @@ func (r UnconnectedReply) Build() bytes.Buffer {
 	outBuffer.Write(r.ID)
 	outBuffer.Write(r.Magic)
 
-	serverNameLen := uint16(len(r.ServerName))
+	pongDataString := writePong(r.Pong)
+	pongDataLen := len(pongDataString)
+
 	stringBuf := make([]byte, 2)
-	binary.BigEndian.PutUint16(stringBuf, serverNameLen)
+	binary.BigEndian.PutUint16(stringBuf, uint16(pongDataLen))
 
 	outBuffer.Write(stringBuf)
-	outBuffer.WriteString(r.ServerName)
+	outBuffer.WriteString(pongDataString)
 
 	return outBuffer
+}
+
+func readPong(raw string) PongData {
+	pong := PongData{}
+	pongParts := []interface{}{}
+
+	stringParts := strings.Split(raw, ";")
+	for _, val := range stringParts {
+		pongParts = append(pongParts, val)
+	}
+
+	util.MapFieldsToStruct(pongParts, &pong)
+
+	fmt.Printf("%v\n", pong)
+	return pong
+}
+
+func writePong(pong PongData) string {
+	var pongDataFields []string
+	pongDataFieldsRaw := util.MapStructToFields(&pong)
+	for _, value := range pongDataFieldsRaw {
+		pongDataFields = append(pongDataFields, fmt.Sprintf("%v", value))
+	}
+
+	return strings.Join(pongDataFields, ";")
 }

--- a/internal/proxy/proxy.go
+++ b/internal/proxy/proxy.go
@@ -6,7 +6,6 @@ import (
 	"math/rand"
 	"net"
 	"os"
-	"strings"
 	"time"
 
 	"github.com/jhead/phantom/internal/proto"
@@ -259,8 +258,6 @@ func (proxy *ProxyServer) processDataFromServer(remoteConn *net.UDPConn, client 
 		if packetID := data[0]; packetID == proto.UnconnectedReplyID {
 			if packet, err := proto.ReadUnconnectedReply(data); err == nil {
 				// Rewrite server MOTD to remove ports
-				truncServerName := strings.Split(packet.ServerName, ";")[:9]
-				packet.ServerName = fmt.Sprintf("%v;", strings.Join(truncServerName, ";"))
 				packetBuffer := packet.Build()
 				data = packetBuffer.Bytes()
 			} else {

--- a/internal/util/struct_iter.go
+++ b/internal/util/struct_iter.go
@@ -1,0 +1,29 @@
+package util
+
+import (
+	"reflect"
+)
+
+func MapFieldsToStruct(fields []interface{}, target interface{}) {
+	numFields := len(fields)
+	targetStruct := reflect.ValueOf(target).Elem()
+
+	for i := 0; i < targetStruct.NumField() && i < numFields; i++ {
+		field := targetStruct.Field(i)
+		fieldValue := reflect.ValueOf(fields[i])
+		field.Set(fieldValue)
+	}
+}
+
+func MapStructToFields(source interface{}) []interface{} {
+	var fields []interface{}
+
+	sourceStruct := reflect.ValueOf(source).Elem()
+
+	for i := 0; i < sourceStruct.NumField(); i++ {
+		field := sourceStruct.Field(i).Interface()
+		fields = append(fields, field)
+	}
+
+	return fields
+}

--- a/internal/util/struct_iter.go
+++ b/internal/util/struct_iter.go
@@ -4,14 +4,22 @@ import (
 	"reflect"
 )
 
+// Takes an arbitrary struct reference and apples all the values from
+// an array to its fields.
 func MapFieldsToStruct(fields []interface{}, target interface{}) {
 	numFields := len(fields)
 	targetStruct := reflect.ValueOf(target).Elem()
 
+	// Iterate over the fields in the struct and array using the same index
 	for i := 0; i < targetStruct.NumField() && i < numFields; i++ {
 		field := targetStruct.Field(i)
 		fieldValue := reflect.ValueOf(fields[i])
-		field.Set(fieldValue)
+
+		// Make sure the types match first, otherwise skip
+		if field.Type() == fieldValue.Type() {
+			// Put the value from the array into the struct
+			field.Set(fieldValue)
+		}
 	}
 }
 

--- a/internal/util/struct_iter_test.go
+++ b/internal/util/struct_iter_test.go
@@ -45,6 +45,15 @@ func TestToStructWithMissing(t *testing.T) {
 	assert.Equal(t, Foo{"a", "b", ""}, foo)
 }
 
+func TestToStructWithPartialMismatch(t *testing.T) {
+	fields := []interface{}{"a", 1, "c"}
+	foo := Foo{}
+
+	MapFieldsToStruct(fields, &foo)
+
+	assert.Equal(t, Foo{"a", "", "c"}, foo)
+}
+
 func TestToFields(t *testing.T) {
 	fields := MapStructToFields(&Foo{"a", "b", "c"})
 	expected := []interface{}{"a", "b", "c"}

--- a/internal/util/struct_iter_test.go
+++ b/internal/util/struct_iter_test.go
@@ -1,0 +1,60 @@
+package util
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type Foo struct {
+	A string
+	B string
+	C string
+}
+
+type Bar struct {
+	A string
+	B int
+	C bool
+}
+
+func TestToStruct(t *testing.T) {
+	fields := []interface{}{"a", "b", "c"}
+	foo := Foo{}
+
+	MapFieldsToStruct(fields, &foo)
+
+	assert.Equal(t, Foo{"a", "b", "c"}, foo)
+}
+
+func TestToStructWithExtra(t *testing.T) {
+	fields := []interface{}{"a", "b", "c", "d"}
+	foo := Foo{}
+
+	MapFieldsToStruct(fields, &foo)
+
+	assert.Equal(t, Foo{"a", "b", "c"}, foo)
+}
+
+func TestToStructWithMissing(t *testing.T) {
+	fields := []interface{}{"a", "b"}
+	foo := Foo{}
+
+	MapFieldsToStruct(fields, &foo)
+
+	assert.Equal(t, Foo{"a", "b", ""}, foo)
+}
+
+func TestToFields(t *testing.T) {
+	fields := MapStructToFields(&Foo{"a", "b", "c"})
+	expected := []interface{}{"a", "b", "c"}
+
+	assert.Equal(t, expected, fields)
+}
+
+func TestToFieldsWithTypes(t *testing.T) {
+	fields := MapStructToFields(&Bar{"a", 1, true})
+	expected := []interface{}{"a", 1, true}
+
+	assert.Equal(t, expected, fields)
+}

